utility/session.go

```diff
@@ -0,0 +1,138 @@
+package utility
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"github.com/pokt-network/pocket/shared/crypto"
+	"github.com/pokt-network/pocket/shared/modules"
+	"github.com/pokt-network/pocket/utility/types"
+)
+
+type Session interface {
+	NewSession(sessionHeight int64, blockHash string, geoZone GeoZone, relayChain RelayChain, application modules.Actor) (Session, types.Error)
+	GetServiceNodes() []modules.Actor // the ServiceNodes providing Web3 to the application
+	GetFishermen() []modules.Actor    // the Fishermen monitoring the serviceNodes
+	GetApplication() modules.Actor    // the Application consuming the web3 access
+	GetRelayChain() RelayChain        // the chain identifier of the web3
+	GetGeoZone() GeoZone              // the geolocation zone where all are registered
+	GetSessionHeight() int64          // the block height when the session started
+}
+
+type RelayChain Identifier
+type GeoZone Identifier
+
+type Identifier interface {
+	Name() string
+	ID() string
+	Bytes() []byte
+}
+
+var _ Session = &session{}
+
+type session struct {
+	serviceNodes  []modules.Actor
+	fishermen     []modules.Actor
+	application   modules.Actor
+	relayChain    RelayChain
+	geoZone       GeoZone
+	blockHash     string
+	key           []byte
+	sessionHeight int64
+}
+
+func (s *session) NewSession(sessionHeight int64, blockHash string, geoZone GeoZone, relayChain RelayChain, application modules.Actor) (session Session, err types.Error) {
+	s.sessionHeight = sessionHeight
+	s.blockHash = blockHash
+	s.geoZone = geoZone
+	s.relayChain = relayChain
+	s.application = application
+	s.key, err = s.sessionKey()
+	if err != nil {
+		return
+	}
+	s.serviceNodes = s.findClosestXServiceNodes()
```

Comment:
`s.serviceNodes = s.findClosestXServiceNodes()`

This function is neither `functional` (i.e. "pure") not `imperative` (i.e. object-oriented).

I think that we should either do one of:

1. `s.serviceNodes = findClosestXServiceNodes(..., list, of, arguments, here)`
2. `err := s.findClosestXServiceNodes()` where the underlying pointer receiver updates the internals of the object (using it's own state) and returns an error if one takes place.

Reply:
I'm not going to be a stickler on this, but though you are correct with `This function is neither functional (i.e. "pure") not imperative (i.e. object-oriented).` it's very human readable and describes exactly what it's doing without you needing to understand the nuances of the underlying language

Reply:
https://github.com/pokt-network/pocket/pull/303#issuecomment-1284187289

Reply:
Might ask to change this in the future (when we implement it) but no need to block.